### PR TITLE
Move servlet instrumentations around

### LINKED ISSUE #586: Move :instrumentation:servlet parent module contents to a submodule
This is the only example where there's a parent module with content _and_ child modules (outside of cases where the child modules are only test modules).

The other prior examples were moved into, e.g. `mongo-common`, but the servlet parent module is slightly different because it has standalone instrumentation.

This would be good to resolve before we publish to maven central.

### PR DESCRIPTION
**Pull request by iNikem**
Closes #586 

But I need help with muzzle. `servlet-common` module has instrumentations which work both for Servlet API 2.3 and 3.0. And it gets included into `servlet-2.3` and `servlet-3.0` modules, which both expect muzzle failure with other API version. But now some instrumentation happily pass both.

### CODE REVIEW

**Comment by trask** [comment]
Yeah, I wasn't sure what to do with the stuff in the servlet root module. It doesn't fit the pattern of other instrumentation. Will look at it tomorrow 👍

**Comment by trask** [comment]
I'm in favor of deleting `FilterInstrumentation` and `HttpServletInstrumentation` (which are disabled by default anyways).

That doesn't really solve our problem though, since we still have to deal with `RequestDispatcherInstrumentation`.

I think it makes sense for `RequestDispatcherInstrumentation` to go in `servlet-2.3` (which actually should be `servlet-2.2` based on looking at it's muzzle definition just now).

Then `servlet-2.2` and `servlet-3.0` would be sort of like `mongo-3.1` and `mongo-3.7`, or `hibernate-4.0` and `hibernate-4.3`, which overlap in the versions that they apply to.

The remaining problem though (and the difference with the mongo and hibernate example) is that one instrumentation in `servlet-2.2` (`Servlet2Instrumentation`) doesn't apply to 3.0+, and one instrumentation (`RequestDispatcherInstrumentation`) does.

And muzzle doesn't like that, or at least it doesn't like it because we now included `classLoaderMatcher` in the muzzle decision (https://github.com/DataDog/dd-trace-java/pull/1155).

That leaves us with the option of copying `RequestDispatcherInstrumentation` and `HttpServletResponseInstrumentation` and their tests to each of `servlet-2.2` and `servlet-3.0`, and applying a `classLoaderMatcher` filter that will apply only to that range for each copy, e.g.

for `servlet-2.2`:

```
  @Override
  public ElementMatcher<ClassLoader> classLoaderMatcher() {
    // Optimization for expensive typeMatcher.
    return hasClassesNamed("javax.servlet.RequestDispatcher")
        .and(not(hasClassesNamed("javax.servlet.AsyncContext")));
  }
```

for `servlet-3.0`:

```
  @Override
  public ElementMatcher<ClassLoader> classLoaderMatcher() {
    // Optimization for expensive typeMatcher.
    return hasClassesNamed("javax.servlet.RequestDispatcher")
        .and(hasClassesNamed("javax.servlet.AsyncContext"));
  }
```

If we don't want to go down the copying route, maybe we could have different `muzzle` specs for different Instrumentations within a given module.

Then we could put Instrumentation in the lowest module version where it applies (in this example, moving everything from the servlet parent to `servlet-2.2`). And more recent versions (e.g. `servlet-3.0`) can layer additional instrumentation on top of that. This would be pretty similar to the `mongo` and `hibernate` examples, except that `servlet-2.2` will have one instrumentation (`Servlet2Instrumentation`) that doesn't apply in 3.0+.

**Comment by iNikem** [comment]
First, let us imagine, that we have 3 modules: `servlet-2.3`, `servlet-3.0` and `servlet-common`. They are independent from each other and contains instrumentations that, correspondingly, apply only to servlet 2.3, only to servlet 3.0 and to both. So far so good.

Now we have 2 classes: `HttpServletRequestGetter` and `ServletHttpServerTracer`, that are needed at least for `servlet-2.3` and `servlet-3.0`. We need to put them somewhere. But as they are not byte-buddy related, they actually can go to that to-be-created "assisted" module, which can safely be included in all auto-instrumentation modules.

So we have to come up with folder structure which includes non-auto instrumentation related code. We have 2 proposals so far. Mine is "assisted", which is probably bad for folder name. Anuraag's is "core", which is Ok in my opinion, but not ideal :) Any ideas?

**Comment by trask** [comment]
I created #621 to discuss what to do with the `servlet-common` module. I suggest we leave it as is in this PR.

**Comment by iNikem** [comment]
> I created #621 to discuss what to do with the `servlet-common` module. I suggest we leave it as is in this PR.

We cannot leave it as is, because this PR fails muzzle

**Comment by iNikem** [comment]
The easy solution, of course, is to drop support for Servlet API 2.X :) Servlet API 3.0 was released 10 years ago. Does anybody use 2.X still?

**Comment by trask** [comment]
> I suggest we leave it as is in this PR.

sorry, that was not clear. I meant leaving `servlet-common` as it is in master (not changing it in this PR).

and deleting FilterInstrumentation and HttpServletInstrumentation (which are disabled by default anyways)

and copying RequestDispatcherInstrumentation and HttpServletResponseInstrumentation and their tests to each of `servlet-2.2` and `servlet-3.0`, and applying a `classLoaderMatcher` filter to restrict each copy, e.g.

for `servlet-2.2`:

```
  @Override
  public ElementMatcher<ClassLoader> classLoaderMatcher() {
    // Optimization for expensive typeMatcher.
    return hasClassesNamed("javax.servlet.RequestDispatcher")
        .and(not(hasClassesNamed("javax.servlet.AsyncContext")));
  }
```

for `servlet-3.0`:

```
  @Override
  public ElementMatcher<ClassLoader> classLoaderMatcher() {
    // Optimization for expensive typeMatcher.
    return hasClassesNamed("javax.servlet.RequestDispatcher")
        .and(hasClassesNamed("javax.servlet.AsyncContext"));
  }
```

**Comment by trask** [comment]
nevermind 😄, I like your proposal from above better https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/601#issuecomment-651594228 👍

**Comment by iNikem** [comment]
@trask ready for review at last

**Comment by iNikem** [comment]
Tests in some instrumentations, such as spring or jetty, depend on servlet instrumentations. They usually depend on `servlet-3.0`. Depending on where some shared instrumentation resides, those tests may or may not produce some extra spans, such as servlet redirects and such. That is way tests change depending on the locations of specific instrumentations.

File: instrumentation/servlet/servlet-2.3/servlet-2.3.gradle [hunk 1]
```diff
@@ -23,8 +23,8 @@ testSets {
 }
 
 dependencies {
-  compileOnly group: 'javax.servlet', name: 'servlet-api', version: '2.2'
-  implementation(project(':instrumentation:servlet:servlet-common'))
+  compileOnly group: 'javax.servlet', name: 'servlet-api', version: '2.3'
```
[the remark this answers is not in the document]
Comment (iNikem) [reply instrumentation/servlet/servlet-2.3/servlet-2.3.gradle:26]: This modules now seems to be fine with `2.2`. 21 years old version... Do you want me to rename the module?

[the remark this answers is not in the document]
Comment (iNikem) [reply instrumentation/servlet/servlet-2.3/servlet-2.3.gradle]: Does not matter any more :) ✅

**Pull request merged**